### ISSUE TITLE
getModuleConfigInputfields() can no longer be static and arguments now need to be InputfieldWrapper instead of array

### ISSUE THREAD

**Issue by adrianbj** [processwire/processwire-issues#997]
Not sure when this happened, but sometime since 3.0.141 you get an exception if you have modules that declare `getModuleConfigInputfields` as a static function. Previously this was ok.

**Comment by teppokoivula** [comment]
@adrianbj, could this be related to https://github.com/processwire/processwire-issues/issues/995? 

At least in that case the problem seems to be that since FieldtypeFile now has a non-static getModuleConfigInputfields() method, any Fieldtype extending FieldtypeFile (or FieldtypeImage which extends FieldtypeFile, etc.) cannot have static getModuleConfigInputfields() method.

**Comment by adrianbj** [comment]
Yeah, I think it is probably the same issue. Note that you also get errors like this:
```
ErrorException: Declaration of FieldtypePDF::getModuleConfigInputfields(array $data) should be compatible with ProcessWire\FieldtypeFile::getModuleConfigInputfields(ProcessWire\InputfieldWrapper $inputfields)
```

which is because authors of modules that extend FieldtypeFile have typically defined their `getModuleConfigInputfields()` method like this:

```
public function getModuleConfigInputfields(array $data)
```
but now the core FieldtypeFile does this:

```
public function getModuleConfigInputfields(InputfieldWrapper $inputfields)
```

so the arguments are of different types.

**Comment by teppokoivula** [comment]
Right – it is the same issue indeed. I was going to write a longer reply to the original issue, but I think it's all pretty clear to Ryan anyway. The tricky part is that no matter how you define this method, you're potentially going to break backwards compatibility to some extent.

One relatively foolproof solution could be implementing the config method in a separate config class instead, though I'm not sure if that would work in this context (?).

**Comment by ryancramerdesign** [comment]
Static config methods are fine, but they are a lot less flexible, especially when it comes to extending module classes. I used them primarily in 2.x but converted all static config methods to non-static in PW 3.x. Earlier in 2.x, static methods were necessary for module config because instantiating a module meant also loading all its assets (JS, CSS, etc.) along with any hooks it attaches, which we didn't want to do when just configuring it. As the Modules class was expanded with new capabilities to limit when happens when a module is loaded, that was no longer necessary, so static config methods likewise are no longer necessary. Though they'll remain for backwards compatibility.

For this particular case, the issue is just that FieldtypeFile added a configuration method where there wasn't one before, and I didn't realize that modules extending it were configurable. I've updated FieldtypeFile to use a hookable version of its method which should not conflict with those modules using a static configuration method. Also, the configuration in FieldtypeFile is intended only for FieldtypeFile and not to be inherited to any descending classes (it doesn't do anything if class isn't FieldtypeFile), so it's no problem that descending modules take over the config. I'm not sure FieldtypeFile even needs the config, so I may just remove it if it continues to cause issues with any descending classes.
